include/genomics/printer.hpp

```diff
@@ -162,66 +195,87 @@ namespace genomics {
+    }
 
     return csvline;
   }
 
   template <class t_wt, uint32_t t_dens, uint32_t t_inv_dens>
   std::string get_csv_lines(const genome_index<t_wt, t_dens, t_inv_dens>& gi,
                             const kmer& k, bool start,
-                            std::vector<std::vector<std::tuple<int64_t, match>>>& off_targets) {
+                            int64_t max_off_targets,
+                            std::vector<std::vector<std::tuple<int64_t, match>>>& off_targets, bool complete) {
     std::string csvlines;
 
+    float cfd_sum = 0.0;
+    std::vector<std::string> off_targets_lines;
     for (uint64_t d = 0; d < off_targets.size(); d++) {
-      for (const auto& off_target : off_targets[d]) {
+      for (int64_t i = 0; i < off_targets[d].size(); i++) {
+        if ((max_off_targets != -1) && (i >= max_off_targets)) break;
+        const auto& off_target = off_targets[d][i];
         int64_t off_target_abs_coords = std::get<0>(off_target);
         match match = std::get<1>(off_target);
-        csvlines += get_csv_line(gi, k, start, match, off_target_abs_coords);
-        csvlines += "\n";
+        std::string csvline = get_csv_line(gi, k, start, match, off_target_abs_coords, complete);
+        if (csvline != "") {
+          off_targets_lines.push_back(csvline);
+          cfd_sum += calculate_cfd(k.sequence, complement(match.sequence), k.pam);
+        }
       }
     }
 
+    // Each off-target will have the same specificity value - add at the end of each line
+    float specificity = 0.0;
+    if (cfd_sum > 0) specificity = 1/cfd_sum;
+    for (auto line: off_targets_lines) {
+      csvlines += line + "," + std::to_string(specificity) + "\n";
+    }
+
     return csvlines;
   }
 
   template <class t_wt, uint32_t t_dens, uint32_t t_inv_dens>
   std::string get_sam_line(const genome_index<t_wt, t_dens, t_inv_dens>& gi,
                            const kmer& k, bool start, int64_t max_off_targets,
-                           std::vector<std::vector<std::tuple<int64_t, match>>>& off_targets) {
+                           std::vector<std::vector<std::tuple<int64_t, match>>>& off_targets,
+                           bool complete) {
     std::string sequence = start ? k.pam + k.sequence : k.sequence + k.pam;
-    std::string samline(k.id);
+    std::string samline(k.id);                                     // SAM:QNAME
 
     samline += "\t";
-    samline += k.dir == direction::positive ? "0" : "16";
-    samline += "\t" + k.chromosome;
+    samline += k.dir == direction::positive ? "0" : "16";          // SAM:FLAG (16 => SEQ being reverse complemented)
+    samline += "\t" + k.chromosome;                                // SAM:RNAME
 
-    samline += "\t" + std::to_string(k.position);
-    samline += "\t100";
-    samline += "\t" + std::to_string(sequence.length()) + "M";
-    samline += "\t*\t0\t0";
+    samline += "\t" + std::to_string(k.position + 1);              // SAM:POS (1-indexed)
```

Comment:
1-indexed position. We never read it internally in our code so no other change is needed in our code to account for this.